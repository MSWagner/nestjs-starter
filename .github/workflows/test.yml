name: Test

on: [push, pull_request]

jobs:
  test:
    name: Test
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2

      - name: Build the docker-compose stack
        run: docker-compose -f 'docker-compose-test.yml' up -d
      - name: Check running containers
        run: docker ps -a
      - name: Check logs
        run: docker logs ananas-dark-backend-postgres-test

      - name: Use Node.js
        uses: actions/setup-node@v1
        with:
          node-version: '10.13.0'
      - name: Install dependencies
        run: yarn
      - name: Build App
        run: yarn build
      - name: Run database migrations
        run: yarn db:migrate:test
      - name: Start tests
        run: yarn test