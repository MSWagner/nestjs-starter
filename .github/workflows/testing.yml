name: Testing

on: [push, pull_request]

jobs:
  test-local:
    name: Testing Local
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2

      - name: Build the docker-compose stack
        run: docker-compose -f 'docker-compose.test.yml' up -d
      - name: Check running containers
        run: docker ps -a
      - name: Check logs
        run: docker logs project-name-postgres-test-standalone

      - name: Use Node.js
        uses: actions/setup-node@v1
        with:
          node-version: '18'
      - name: Install dependencies
        run: yarn
      - name: Build App
        run: yarn build
      - name: Run database migrations
        run: yarn db:migrate:test
      - name: Start tests
        run: yarn test

  test-docker:
    name: Testing Docker Container
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2

      - name: Use Node.js
        uses: actions/setup-node@v1
        with:
          node-version: '18'
      
      - name: Create docker network
        run: docker network create project-name-network-dev
      - name: Start Docker Dev Container
        run: yarn docker:up:hidden
      - name: Check running containers
        run: docker ps -a
      - name: Check logs
        run: docker logs project-name-postgres-test

      - name: Run database migrations
        run: yarn docker:db:migrate

      - name: Start tests in docker container
        run: yarn docker:test