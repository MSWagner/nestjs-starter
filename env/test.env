NODE_ENV=test

PGHOST=localhost
PGPORT=5431
PGUSER=root
PGPASSWORD=toor
PGDATABASE=project-name-test-db