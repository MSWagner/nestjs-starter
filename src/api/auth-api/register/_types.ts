import { IsString, MinLength } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";

export class RegisterDto {
    @IsString()
    @MinLength(3)
    @ApiProperty({ minLength: 3 })
    readonly username: string;

    @IsString()
    @MinLength(6)
    @ApiProperty({ minLength: 6 })
    readonly password: string;
}

export class RegisterResponse {
    @ApiProperty()
    readonly tokenType: string;

    @ApiProperty()
    readonly expiresIn: number;

    @ApiProperty({ format: "uuid" })
    refreshToken: string;

    @ApiProperty({ format: "uuid" })
    accessToken: string;
}
