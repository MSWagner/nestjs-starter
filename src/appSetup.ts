import compression from "compression";
import helmet from "helmet";
import rateLimit from "express-rate-limit";

import { INestApplication, ValidationPipe, VersioningType } from "@nestjs/common";

export async function setupApp(app: INestApplication): Promise<void> {
    app.setGlobalPrefix("api");
    app.enableVersioning({
        type: VersioningType.URI
    });

    app.use(compression());
    app.use(helmet());
    app.enableCors();
    app.use(
        rateLimit({
            windowMs: 15 * 60 * 1000, // 15 minutes
            max: 100 // limit each IP to 100 requests per windowMs
        })
    );

    app.useGlobalPipes(new ValidationPipe({ whitelist: true, forbidNonWhitelisted: true }));
}
