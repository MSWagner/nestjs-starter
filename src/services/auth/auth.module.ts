import { Module } from "@nestjs/common";

import { AuthService } from "./auth.service";

import { LocalStrategy } from "./strategies/local.strategy";
import { BearerStrategy } from "./strategies/bearer.strategy";

import { UserModule } from "../user/user.module";

import { PassportModule } from "@nestjs/passport";

import { accessTokenProviders } from "./providers/accessTokenProviders";
import { refreshTokenProviders } from "./providers/refreshTokenProviders";

@Module({
    imports: [UserModule, PassportModule],
    providers: [AuthService, LocalStrategy, BearerStrategy, ...accessTokenProviders, ...refreshTokenProviders],
    exports: [AuthService]
})
export class AuthModule {}
