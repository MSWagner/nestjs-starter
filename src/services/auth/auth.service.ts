import * as bcrypt from "bcrypt";
import moment from "moment";
import * as _ from "lodash";

import { Injectable, Inject } from "@nestjs/common";
import { Repository } from "typeorm";

import { UserService } from "../user/user.service";

import { User } from "../../entities/User.entity";
import { AccessToken } from "../../entities/AccessToken.entity";
import { RefreshToken } from "../../entities/RefreshToken.entity";

import { CONFIG } from "../../config";

export interface IToken {
    accessToken: AccessToken;
    refreshToken: RefreshToken;
}

@Injectable()
export class AuthService {
    constructor(
        @Inject(CONFIG.database.defaultAccessTokenRepoName)
        private readonly accessTokenRepository: Repository<AccessToken>,

        @Inject(CONFIG.database.defaultRefreshTokenRepoName)
        private readonly refreshTokenRepository: Repository<RefreshToken>,

        private readonly userService: UserService
    ) {}

    async generateToken(user: User): Promise<{ accessToken: AccessToken; refreshToken: RefreshToken }> {
        const newRefreshToken = new RefreshToken();
        newRefreshToken.validUntil =
            CONFIG.auth.refreshTokenValidityMS > 0
                ? moment().add(CONFIG.auth.refreshTokenValidityMS, "milliseconds").toDate()
                : null;
        newRefreshToken.user = user;

        const accessToken = await this.generateAuthToken(user);
        const refreshToken = await this.refreshTokenRepository.save(newRefreshToken);

        return {
            accessToken,
            refreshToken
        };
    }

    async generateAuthToken(user: User): Promise<AccessToken> {
        const newAccessToken = new AccessToken();
        newAccessToken.validUntil = moment().add(CONFIG.auth.tokenValidity, "milliseconds").toDate();
        newAccessToken.user = user;

        const accessToken = await this.accessTokenRepository.save(newAccessToken);

        return accessToken;
    }

    async validateUser(username: string, password: string): Promise<User> {
        const user = await this.userService.findOne(username);

        if (_.isNil(user)) {
            return null;
        }

        if (!user.isActive) {
            return null;
        }

        const isValid = await bcrypt.compare(password, user.passwordHash);

        return isValid ? user : null;
    }

    async validateToken(token: string): Promise<User> {
        const accessToken = await this.accessTokenRepository.findOne(token, { relations: ["user"] });

        if (_.isNil(accessToken)) {
            return null;
        }

        const isValid = moment().isSameOrBefore(accessToken.validUntil);

        if (!isValid) {
            await this.accessTokenRepository.delete(accessToken);
        }

        if (!accessToken.user.isActive) {
            return null;
        }

        return isValid ? accessToken.user : null;
    }

    async register(username: string, password: string): Promise<User> {
        return this.userService.createUser(username, password);
    }

    async login(user: User): Promise<IToken> {
        return this.generateToken(user);
    }

    async refreshAuthToken(token: string): Promise<IToken> {
        const refreshToken = await this.refreshTokenRepository.findOne(token, { relations: ["user"] });

        const isValid =
            !_.isNil(refreshToken) &&
            !_.isNil(refreshToken.user) &&
            (_.isNil(refreshToken.validUntil) || moment().isSameOrBefore(refreshToken.validUntil));

        if (!isValid && !_.isNil(refreshToken)) {
            await this.refreshTokenRepository.delete(refreshToken);
        }

        return !isValid
            ? null
            : {
                  accessToken: await this.generateAuthToken(refreshToken.user),
                  refreshToken
              };
    }
}
