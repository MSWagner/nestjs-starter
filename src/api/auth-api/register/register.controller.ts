import { Controller, Body, Post, HttpException, HttpStatus } from "@nestjs/common";
import { ApiTags, ApiCreatedResponse } from "@nestjs/swagger";

import { AuthService } from "../../../services/auth/auth.service";
import { RegisterDto, RegisterResponse } from "./_types";

import { CONFIG } from "../../../config";

@ApiTags("auth")
@Controller({ path: "auth/register", version: "1" })
export class RegisterController {
    constructor(private readonly authService: AuthService) {}

    @Post()
    @ApiCreatedResponse({ description: "The user has been successfully created.", type: RegisterResponse })
    async handler(@Body() requestDto: RegisterDto): Promise<RegisterResponse> {
        try {
            const newUser = await this.authService.register(requestDto.username, requestDto.password);
            const tokens = await this.authService.generateToken(newUser);

            return {
                tokenType: "Bearer",
                expiresIn: CONFIG.auth.tokenValidity,
                accessToken: tokens.accessToken.token,
                refreshToken: tokens.refreshToken.token
            };
        } catch (err) {
            throw new HttpException("User with the given username already exists", HttpStatus.CONFLICT);
        }
    }
}
