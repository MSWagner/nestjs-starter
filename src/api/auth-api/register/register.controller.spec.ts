import request from "supertest";
import { Test, TestingModule } from "@nestjs/testing";
import { INestApplication } from "@nestjs/common";

import { setupApp } from "../../../appSetup";

import { TestModule } from "../../../services/test/test.module";
import { TestService } from "../../../services/test/test.service";
import { RegisterController } from "./register.controller";
import { AuthModule } from "../../../services/auth/auth.module";

import { User } from "../../../entities/User.entity";

import * as fixtures from "../../../services/test/fixtures";

describe("Register Controller", () => {
    let app: INestApplication;
    let testService: TestService;

    beforeAll(async () => {
        const module: TestingModule = await Test.createTestingModule({
            controllers: [RegisterController],
            imports: [TestModule, AuthModule]
        }).compile();

        testService = module.get<TestService>(TestService);

        app = module.createNestApplication();
        setupApp(app);
        await app.init();
    });

    beforeEach(async () => {
        await testService.reloadFixtures();
    });

    afterAll((done) => {
        testService.connectionManager.connection.close();
        done();
    });

    it("/POST auth/register - should register an new user", async () => {
        const registerBody = {
            username: "test1",
            password: "passwordTest"
        };

        const userCountBefore = await User.count({ username: registerBody.username });
        expect(userCountBefore).toEqual(0);

        const response = await request(app.getHttpServer())
            .post("/api/v1/auth/register")
            .send(registerBody)
            .expect(201);

        const userCountAfter = await User.count({ username: registerBody.username });
        expect(userCountAfter).toEqual(1);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterCredentials");
    });

    it("/POST auth/register - should throw an error for a username conflict", async () => {
        const registerBody = {
            username: fixtures.user1.username,
            password: "passwordTest"
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/auth/register")
            .send(registerBody)
            .expect(409);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterUsernameConflictError");
    });

    it("/POST auth/register - should throw an error for missing username", async () => {
        const registerBody = {
            password: "passwordTest"
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/auth/register")
            .send(registerBody)
            .expect(400);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterUsernameMissingError");
    });

    it("/POST auth/register - should throw an error for missing password", async () => {
        const registerBody = {
            username: "test1"
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/auth/register")
            .send(registerBody)
            .expect(400);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterPasswordMissingError");
    });

    it("/POST auth/register - should throw an error for a to short password", async () => {
        const registerBody = {
            username: "test1",
            password: "passw"
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/auth/register")
            .send(registerBody)
            .expect(400);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterPasswordToShortError");
    });

    it("/POST auth/register - should throw an error for a to short username", async () => {
        const registerBody = {
            username: "te",
            password: "password"
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/auth/register")
            .send(registerBody)
            .expect(400);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterUsernameToShortError");
    });
});
