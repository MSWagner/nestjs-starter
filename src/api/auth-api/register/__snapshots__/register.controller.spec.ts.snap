// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Controller /POST auth/register - should register an new user: RegisterCredentials 1`] = `
{
  "accessToken": "accessToken",
  "expiresIn": 86400000,
  "refreshToken": "refreshToken",
  "tokenType": "Bearer",
}
`;

exports[`Register Controller /POST auth/register - should throw an error for a to short password: RegisterPasswordToShortError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "password must be longer than or equal to 6 characters",
  ],
  "statusCode": 400,
}
`;

exports[`Register Controller /POST auth/register - should throw an error for a to short username: RegisterUsernameToShortError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "username must be longer than or equal to 3 characters",
  ],
  "statusCode": 400,
}
`;

exports[`Register Controller /POST auth/register - should throw an error for a username conflict: RegisterUsernameConflictError 1`] = `
{
  "message": "User with the given username already exists",
  "statusCode": 409,
}
`;

exports[`Register Controller /POST auth/register - should throw an error for missing password: RegisterPasswordMissingError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "password must be longer than or equal to 6 characters",
    "password must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`Register Controller /POST auth/register - should throw an error for missing username: RegisterUsernameMissingError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "username must be longer than or equal to 3 characters",
    "username must be a string",
  ],
  "statusCode": 400,
}
`;
